File: formula-java/src/main/java/org/formula/parse/tree/MatcherNode.java
package org.formula.parse.tree;

import java.util.Objects;
import java.util.stream.Stream;
import org.formula.util.StreamUtils;

public class MatcherNode<T> extends MappableNode<T> {
    protected final TokenMatcher matcher;

    @Override
    public Stream<TokenMatch<T>> walk(String text, int startIndex, int currentIndex) {
        if (!matches(text, startIndex, currentIndex)) {
            return Stream.empty();
        }
        var matches = walkChildren(text, startIndex, currentIndex + 1);
        if (mapper == null) {
            return matches;
        }
        var match = new TokenMatch<>(text, startIndex, currentIndex + 1, mapper);
        return StreamUtils.defaultIfEmpty(matches,
                () -> match);
    }

    protected boolean matches(String text, int startIndex, int currentIndex) {
        if (currentIndex >= text.length()) {
            return false;
        }
        char current = text.charAt(currentIndex);
        return matcher.matches(current);
    }

    @Override
    public boolean equals(Object other) {
        if (other == null) {
            return false;
        }
        return getClass() == other.getClass()
                && Objects.equals(matcher, ((MatcherNode<T>) other).matcher);
    }

    public MatcherNode(TokenMatcher matcher) {
        this.matcher = matcher;
    }
}

File: cs/formula-cs/Formula/TokenTree/MatcherNode.cs
namespace Formula.TokenTree;

internal class MatcherNode<T> : MappableNode<T>
{
    protected readonly ITokenMatcher Matcher;

    public MatcherNode(ITokenMatcher matcher)
    {
        Matcher = matcher;
    }

    public override void Walk(string text, int startIndex, int currentIndex, List<TokenMatch<T>> matches)
    {
        if (!Matches(text, startIndex, currentIndex))
        {
            return;
        }
        
        WalkChildren(text, startIndex, currentIndex + 1, matches);
        if (Mapper == null)
        {
            return;
        }
        var match = new TokenMatch<T>(text, startIndex, currentIndex + 1, Mapper);
        matches.Add(match);
    }
    
    protected bool Matches(string text, int startIndex, int currentIndex) {
        if (currentIndex >= text.Length) {
            return false;
        }
        var current = text[currentIndex];
        return Matcher.Matches(current);
    }
    
    public override bool Equals(object? other)
    {
        return other is MatcherNode<T> node && Equals(node);
    }

    private bool Equals(MatcherNode<T> other)
    {
        return Matcher.Equals(other.Matcher);
    }

    public override int GetHashCode()
    {
        return Matcher.GetHashCode();
    }
}